src/main/java/org/opensearch/knn/index/query/nativelib/NativeEngineKnnVectorQuery.java

```diff
@@ -76,101 +77,131 @@ public Weight createWeight(IndexSearcher indexSearcher, ScoreMode scoreMode, flo
             log.debug("Rescoring results took {} ms. oversampled k:{}, segments:{}", rescoreTime, firstPassK, leafReaderContexts.size());
         }
         ResultUtil.reduceToTopK(perLeafResults, finalK);
+
+        if (expandNestedDocs) {
+            perLeafResults = retrieveAll(indexSearcher, leafReaderContexts, knnWeight, perLeafResults, rescoreContext == null);
+        }
```

Comment:
This code is not exclusively for rescoring; it will be executed even when there is no rescore context. Refer to the changes in `KNNQueryFactory` at line #114.
```
            } else if (ENGINES_SUPPORTING_MULTI_VECTORS.contains(knnEngine) && expandNested) {
                return new NativeEngineKnnVectorQuery(knnQuery, QueryUtils.INSTANCE, expandNested);
```